File: src/main/scala/nl/jqno/remindermail/ConfigActivity.scala
/*
 * ReminderMail is an Android app that will let you quickly send e-mail to
 * a pre-defined address.
 *
 * Copyright (C) 2011 by Jan Ouwens
 *
 * This file is part of ReminderMail.
 *
 * ReminderMail is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * ReminderMail is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with ReminderMail.  If not, see <http://www.gnu.org/licenses/>.
 */
package nl.jqno.remindermail

import android.app.Activity
import android.content.Intent
import android.database.Cursor
import android.net.Uri
import android.os.Bundle
import android.view.ContextMenu
import android.view.ContextMenu.ContextMenuInfo
import android.view.MenuItem
import android.view.View
import android.widget._

import FindView._

class ConfigActivity extends Activity with FindView {
  private val PickContact = 1337
  private val ColorLines = List(
    R.id.config_step1_line,
    R.id.config_step2_line,
    R.id.config_step3_line,
    R.id.config_step4_line,
    R.id.config_done_line
  )

  private lazy val state = new State(this)
  private lazy val contextMenuView = find(R.id.config_invisible_context_menu_view)

  private var nameCandidate = ""
  private var emailCandidates = List.empty[String]

  override def onCreate(savedInstanceState: Bundle): Unit = {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.config)
    findView[Button](R.id.config_step2_go) onClick selectMail()
    findView[Button](R.id.config_step3_clear) onClick clearTag()
    findView[Button](R.id.config_done) onClick finish()
    findView[Button](R.id.config_about) onClick about()

    val edit = findView[EditText](R.id.config_step3_tag)
    edit.setSelection(edit.getText.length)
    edit.setText(state.tag)
    edit onTextChanged tagChanged()

    findView[ScrollView](R.id.config_scroller).fullScroll(View.FOCUS_UP)
    registerForContextMenu(contextMenuView)

    paintColorLines()
    paintMail()
  }

  private def selectMail(): Unit = {
    val intent = new Intent(Intent.ACTION_PICK, Contacts.CONTENT_URI)
    startActivityForResult(intent, PickContact)
  }

  private def about(): Unit = {
    val intent = new Intent(this, classOf[AboutActivity])
    startActivity(intent)
  }

  private def clearTag(): Unit = {
    findView[EditText](R.id.config_step3_tag).setText("")
    state.setTag("")
  }

  private def tagChanged(): Unit = {
    val edit = findView[EditText](R.id.config_step3_tag)
    val tag = edit.getText.toString
    state.setTag(tag)
  }

  override def onActivityResult(reqCode: Int, resultCode: Int, data: Intent): Unit = {
    super.onActivityResult(reqCode, resultCode, data)

    reqCode match {
      case PickContact =>
        if (resultCode == Activity.RESULT_OK) {
          updateSettings(data.getData)
        }
    }
  }

  private def updateSettings(contact: Uri): Unit = {
    val contactCursor = managedQuery(contact, null, null, null, null)
    if (contactCursor.moveToFirst()) {
      val id = getString(contactCursor, Contacts.LOOKUP_KEY)

      nameCandidate = getString(contactCursor, Contacts.DISPLAY_NAME)
      emailCandidates = List.empty[String]

      val mailCursor = managedQuery(Contacts.EMAIL_CONTENT_URI, null, s"${Contacts.EMAIL_CONTACT_ID} = ?", Array(id), null)
      while (mailCursor.moveToNext()) {
        val e = getString(mailCursor, Contacts.EMAIL_DATA)
        emailCandidates = e :: emailCandidates
      }

      if (emailCandidates.size == 1) {
        state.setNameAndMail(nameCandidate, emailCandidates(0))
        paintMail()
      }
      else {
        openContextMenu(contextMenuView)
      }
    }
  }

  private def getString(cursor: Cursor, id: String): String =
    cursor.getString(cursor getColumnIndex id)

  override def onCreateContextMenu(menu: ContextMenu, v: View, menuInfo: ContextMenuInfo): Unit = {
    super.onCreateContextMenu(menu, v, menuInfo)
    menu.setHeaderTitle(R.string.config_step2_pick_mail)
    for ((email, index) <- emailCandidates.zipWithIndex) {
      menu.add(0, index, 0, email)
    }
  }

  override def onContextItemSelected(item: MenuItem): Boolean = {
    state.setNameAndMail(nameCandidate, emailCandidates(item.getItemId))
    paintMail()
    return true
  }

  private def paintColorLines(): Unit =
    for (line <- ColorLines) { find(line).setBackgroundResource(state.color) }

  private def paintMail(): Unit =
    if (state.mail.isDefined)
      showMail()
    else
      hideMail()

  private def showMail(): Unit = {
    findView[TextView](R.id.config_step2_name).setText(state.name.get)
    findView[TextView](R.id.config_step2_mail).setText(state.mail.get)
    findView[TextView](R.id.config_step2_empty).setVisibility(View.GONE)
    findView[TableLayout](R.id.config_step2_setting_box).setVisibility(View.VISIBLE)
  }

  private def hideMail(): Unit = {
    findView[TextView](R.id.config_step2_name).setText("")
    findView[TextView](R.id.config_step2_mail).setText("")
    findView[TextView](R.id.config_step2_empty).setVisibility(View.VISIBLE)
    findView[TableLayout](R.id.config_step2_setting_box).setVisibility(View.GONE)
  }
}

File: app/src/main/java/nl/jqno/remindermail/ConfigActivity.kt
package nl.jqno.remindermail

import android.Manifest
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.database.Cursor
import android.net.Uri
import android.os.Bundle
import android.provider.ContactsContract
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatActivity
import android.text.Editable
import android.text.TextWatcher
import android.view.ContextMenu
import android.view.MenuItem
import android.view.View
import android.view.WindowManager
import android.widget.Toast
import kotlinx.android.synthetic.main.activity_config.*

const val pickContact = 1337

class ConfigActivity : AppCompatActivity() {

    private val state by lazy { State(this) }
    private var nameCandidate = ""
    private var emailCandidates = emptyList<String>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_config)
        window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_HIDDEN)
        config_scroller.fullScroll(View.FOCUS_UP)
        registerForContextMenu(config_invisible_context_menu)

        step2_go.setOnClickListener { selectMail() }
        step3_clear.setOnClickListener { clearTag() }
        done.setOnClickListener { finish() }
        about.setOnClickListener { openAbout() }

        step3_tag.setText(state.tag())
        step3_tag.setSelection(step3_tag.text.length)
        step3_tag.addTextChangedListener(object: TextWatcher {
            override fun afterTextChanged(s: Editable?) {
                tagChanged()
            }
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
        })

        paintMail()
    }

    private fun selectMail() {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.READ_CONTACTS) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.READ_CONTACTS), pickContact)
        }
        val intent = Intent(Intent.ACTION_PICK, ContactsContract.Contacts.CONTENT_URI)
        startActivityForResult(intent, pickContact)
    }

    private fun clearTag() {
        step3_tag.setText("")
        state.setTag("")
    }

    private fun openAbout() {
        val intent = Intent(this, AboutActivity::class.java)
        startActivity(intent)
    }

    private fun tagChanged() {
        val tag = step3_tag.text.toString()
        state.setTag(tag)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == pickContact && resultCode == Activity.RESULT_OK)
            data?.data?.let { updateSettings(it) }
    }

    private fun updateSettings(contact: Uri) {
        val cr = contentResolver
        val cursor = cr.query(contact, null, null, null, null)
        cursor?.use { cur ->
            if (cur.moveToFirst()) {
                nameCandidate = getString(cur, ContactsContract.Contacts.DISPLAY_NAME)
                emailCandidates = emptyList()

                val id = getString(cur, ContactsContract.Contacts._ID)
                val mailCursor = cr.query(ContactsContract.CommonDataKinds.Email.CONTENT_URI, null,
                        "${ContactsContract.CommonDataKinds.Email.CONTACT_ID} = ?", arrayOf(id), null)
                mailCursor?.use { mc ->
                    while (mc.moveToNext()) {
                        val e = getString(mc, ContactsContract.CommonDataKinds.Email.DATA)
                        emailCandidates += e
                    }
                }

                when (emailCandidates.size) {
                    0 -> {
                        Toast.makeText(this, "No e-mail address found for this contact", Toast.LENGTH_SHORT).show()
                    }
                    1 -> {
                        state.setNameAndMail(nameCandidate, emailCandidates[0])
                        paintMail()
                    }
                    else -> {
                        openContextMenu(config_invisible_context_menu)
                    }
                }
            }
        }
    }

    private fun getString(cursor: Cursor, id: String): String =
            cursor.getString(cursor.getColumnIndex(id))

    private fun paintMail() {
        if (state.mail() == null) {
            step2_name.text = ""
            step2_email.text = ""
            step2_empty.visibility = View.VISIBLE
            step2_settingbox.visibility = View.GONE
        }
        else {
            step2_name.text = state.name()!!
            step2_email.text = state.mail()!!
            step2_empty.visibility = View.GONE
            step2_settingbox.visibility = View.VISIBLE
        }
    }

    override fun onCreateContextMenu(menu: ContextMenu?, v: View?, menuInfo: ContextMenu.ContextMenuInfo?) {
        super.onCreateContextMenu(menu, v, menuInfo)
        menu?.setHeaderTitle(R.string.config_step2_pick_mail)
        emailCandidates.withIndex().forEach {
            menu?.add(0, it.index, 0, it.value)
        }
    }

    override fun onContextItemSelected(item: MenuItem?): Boolean {
        state.setNameAndMail(nameCandidate, emailCandidates[item!!.itemId])
        paintMail()
        return true
    }
}
